onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc

```diff
@@ -433,25 +434,187 @@ Status UpdateFeeds(
+}
+
+// Copy present state to past state for T5 model
+template <typename T>
+void PickT5PastState(const std::vector<OrtValue>& last_outputs,
+                     std::vector<OrtValue>& next_inputs,
+                     int num_present_tensors,
+                     gsl::span<const int32_t>& beam_indices,
+                     AllocatorPtr allocator) {
+  for (int i = 0; i < num_present_tensors; ++i) {
+    const OrtValue& present = last_outputs[transformers::T5DecoderSubgraph::kFirstPresentOutputIndex + i];
+
+    // shape is like (batch_beam_size, 12, past_seq_len, 64)
+    const TensorShape& past_shape = present.Get<Tensor>().Shape();
+    auto block_size_per_beam = past_shape[1] * past_shape[2] * past_shape[3];
+
+    // Create a tensor with same shape.
+    // TODO(tianleiwu): allocate one buffer for all layers
+    OrtValue past;
+    Tensor::InitOrtValue(DataTypeImpl::GetType<T>(), past_shape, allocator, past);
+
+    gsl::span<T> past_span = gsl::make_span<T>(past.GetMutable<Tensor>()->MutableData<T>(), past_shape.Size());
+    gsl::span<const T> present_span = gsl::make_span<const T>(present.Get<Tensor>().Data<T>(), past_shape.Size());
+    for (gsl::index j = 0; j < beam_indices.length(); j++) {
+      int32_t beam_index = beam_indices[j];
+      gsl::span<const T> present_beam = present_span.subspan(beam_index * block_size_per_beam, block_size_per_beam);
+      gsl::span<T> past_beam = past_span.subspan(j * block_size_per_beam, block_size_per_beam);
+      gsl::copy(present_beam, past_beam);
+    }
+
+    next_inputs[transformers::T5DecoderSubgraph::kFirstPastInputIndex + i] = past;
+  }
+}
+
+// Update decoder inputs given decoder outputs of last iteration.
+template <typename T>
+Status UpdateDecoderFeeds(
+    AllocatorPtr allocator,
+    void* stream,
+    const std::vector<OrtValue>& last_outputs,
+    std::vector<OrtValue>& next_inputs,
+    int num_present_tensors,
+    gsl::span<const int32_t> beam_next_tokens,
+    gsl::span<const int32_t> beam_indices,
+    int num_beams,
+    const transformers::IConsoleDumper* dumper) {
+  ORT_UNUSED_PARAMETER(stream);
+
+  // last_outputs: logits, present_key_self_0, present_value_self_0, ...
+  // next_inputs: input_ids,
+  //              encoder_attention_mask, encoder_hidden_states,
+  //              past_key_self_0, past_value_self_0, ...
+  //              past_key_cross_0, past_value_cross_0, ...
+  // Only need copy beam next tokens to input_ids, and copy present_*_self_* to past_*_self_*,
+
+  // Update input_ids with next tokens.
+  int batch_beam_size = static_cast<int>(beam_next_tokens.length());
+  int64_t dims[] = {batch_beam_size, 1};
+  TensorShape input_ids_shape(&dims[0], 2);
+
+  // TODO(tianleiwu): Reuse buffer for input_ids to reduce memory allocation.
+  OrtValue input_ids;
+  Tensor::InitOrtValue(DataTypeImpl::GetType<int32_t>(), input_ids_shape, allocator, input_ids);
+
+  gsl::copy(beam_next_tokens, input_ids.GetMutable<Tensor>()->MutableDataAsSpan<int32_t>());
```

Comment:
I think here it needs to pass sequence[:, -1] instead of beam_next_token?

Reply:
In BeamSearchBase<T>::GenerateNextToken, the last token of sequence is also copied from beam_next_token:
```
cpu_state.sequences.AppendNextTokenToSequences(beam_indices, beam_next_tokens);
```

So logically, sequence[:,-1] is same as beam_next_token, except that sequence is not contiguous array (sequence has fixed max_seq_length length), so we cannot directly copy from it.